lib/pause_2017/PAUSE/Web/Controller/User/Mfa.pm

```diff
@@ -0,0 +1,80 @@
+package PAUSE::Web::Controller::User::Mfa;
+
+use Mojo::Base "Mojolicious::Controller";
+use Auth::GoogleAuth;
+use PAUSE::Crypt;
+use Crypt::URandom qw(urandom);
+use Convert::Base32 qw(encode_base32);
+
+sub edit {
+  my $c = shift;
+  my $pause = $c->stash(".pause");
+  my $mgr = $c->app->pause;
+  my $req = $c->req;
+  my $u = $c->active_user_record;
+
+  my $auth = $c->app->pause->authenticator_for($u);
+  $pause->{mfa_qrcode} = $auth->qr_code;
+  if (!$u->{mfa_secret32}) {
+    my $dbh = $mgr->authen_connect;
+    my $tbl = $PAUSE::Config->{AUTHEN_USER_TABLE};
+    my $sql = "UPDATE $tbl SET mfa_secret32 = ?, changed = ?, changedby = ? WHERE user = ?";
+    $dbh->do($sql, undef, $auth->secret32, time, $pause->{User}{userid}, $u->{userid})
+      or push @{$pause->{ERROR}}, sprintf(qq{Could not enter the data into the database: <i>%s</i>.},$dbh->errstr);
+  }
+
+  if (uc $req->method eq 'POST' and $req->param("pause99_mfa_sub")) {
```

Comment:
This block of code says:

1. The user is updating their mfa
2. If they provided a 6 digit code? Does it verify? No? Fail
3. Did they provide a backup code? Does it verify? No? Fail

This means that you can supply an empty string and successfully disable mfa

It would probably be better to invert the logic and start with `my $success = 0;` and then only set `$success` to true under very explicit circumstances. Then, we only take action if `$success` is true. This can help avoid these kind of logic bugs

Reply:
Inverted the logic with https://github.com/andk/pause/pull/455/commits/684f2f05c7474d86c4d31f7736ee9b0bcb712a60